numpy/_core/numeric.py

```diff
@@ -1121,6 +1121,53 @@ def tensordot(a, b, axes=2):
     return res.reshape(olda + oldb)
 
 
+def _vecdot_dispatcher(x1, x2, /, *, axis=None):
+    return (x1, x2)
+
+
+@array_function_dispatch(_vecdot_dispatcher)
+def vecdot(x1, x2, /, *, axis=-1):
+    """
+    Computes the (vector) dot product of two arrays.
+
+    This function is Array API compatible.
+
+    Parameters
+    ----------
+    x1 : array_like
+        First input array.
+    x2 : array_like
+        Second input array.
+    axis : int, optional
+        Axis over which to compute the dot product. Default: ``-1``.
+
+    Returns
+    -------
+    output : ndarray
+        The vector dot product of the input.
+
+    See Also
+    --------
+    dot
+
+    """
+    ndim = builtins.max(x1.ndim, x2.ndim)
+    x1_shape = (1,)*(ndim - x1.ndim) + tuple(x1.shape)
+    x2_shape = (1,)*(ndim - x2.ndim) + tuple(x2.shape)
+    if x1_shape[axis] != x2_shape[axis]:
+        raise ValueError(
+            "x1 and x2 must have the same size along the dot-compute axis "
+            f"but they are: {x1_shape[axis]} and {x2_shape[axis]}."
+        )
+
+    x1_, x2_ = np.broadcast_arrays(x1, x2)
+    x1_ = np.moveaxis(x1_, axis, -1)
+    x2_ = np.moveaxis(x2_, axis, -1)
+
+    res = x1_[..., None, :] @ x2_[..., None]
```

Comment:
@seberg mentioned on the triage call today that this is supposed to do the hermitian conjugate according to the array API standard - does the matmul operator do that?

Reply:
And if so, probably worth adding a test that has complex vectors.

Reply:
Right, a conjugate was missing - I added it (and a test) according to Array API.

Note: `np.vdot` does conjugate on the first parameter (if it's complex) where `np.vecdot` does it on the second one (I mentioned it in the `vecdot`'s docs).

Reply:
no, the numpy way is correct, the array-api way is just a bug.

Reply:
Ah, right (I see that pytorch also performs conjugate on the first argument). Let me change that here and fix it in the Array API.

Reply:
Note that wikipedia disagrees: the conjugate of the second argument is taken: https://en.wikipedia.org/wiki/Dot_product#Complex_vectors

Reply:
Mathematicians and physicists have opposite conventions. Mathematicians use sesquilinear forms, physicists follow the Dirac bracket convention (I don't know who actually started the convention). It can make teaching awkward ...

Reply:
But then BLAS agrees with what you have, so I guess wikipedia is confused too... https://netlib.org/lapack/explore-html/d1/dcc/group__dot_ga5c189335a4e6130a2206c190579b1571.html#ga5c189335a4e6130a2206c190579b1571